### PR TITLE
feat: Added proper per shard bandwidth metric calculation

### PR DESCRIPTION
**Pull request by NagyZoltanPeter**
# Description
Calculation of relay network traffic per served shards for better understanding network usage for operators.

As part of this, DST log analytics support logs are also added as the necessary solution must converge.
Therefore solution implemented in https://github.com/waku-org/nwaku/pull/2832 is applied here.
This helps testing with lite-protocol-tester and waku-simulator.

# Changes

- [X] Changed rate limit metrics for dashboard
- [X] Updated monitoring dashboard for bw and rate metrics
- [X] proper per shard bandwidth metric calculation
- [X] logging of in/out messages

## related changes in monitoring for nwaku-compose
https://github.com/waku-org/nwaku-compose/pull/99

### Monitoring panel added for relay/shard metrics:
image

## Issue
https://github.com/waku-org/nwaku/issues/1945

### CODE REVIEW

**Review by Ivansete-status — COMMENTED** [review]
Great PR! Thanks for it! You are solving a very complex issue wrt how to properly track the "sent" messages :partying_face: 

I'm not approving yet because we need to comment a little what is the best approach to track the "in" traffic.

If we merge as it is now we will get metrics of all the gross "in" traffic. In other words, we would track all the "in" traffic and even for the topics "my peer" are not interested in or messages that don't bypass the msg validations. I think this metric sounds interesting, to see how loaded "my peer" is, but I think that is something we can add in a separate PR.

On the other hand, I suggest having the "in" traffic metrics logic within the following `else` statement so that we can print the valid messages "my peer" is interested in: https://github.com/waku-org/nwaku/blob/99149ea9dc252c8d6b6d563d0804c1c474a528db/waku/waku_relay/protocol.nim#L273-L274

----

As an additional context, the current approach is tracking the "in" traffic from: https://github.com/vacp2p/nim-libp2p/blob/d0af3fbe8559f69195657a360c3dd4ac4552c811/libp2p/protocols/pubsub/gossipsub.nim#L498

Whereas the suggested approach will track "in" traffic from: https://github.com/vacp2p/nim-libp2p/blob/d0af3fbe8559f69195657a360c3dd4ac4552c811/libp2p/protocols/pubsub/gossipsub.nim#L445

File: apps/liteprotocoltester/monitoring/configuration/dashboards/nwaku-monitoring.json [hunk 1]
```diff
@@ -88,7 +88,7 @@
         "sizing": "auto",
         "valueMode": "color"
       },
-      "pluginVersion": "10.2.3",
+      "pluginVersion": "10.4.2",
```
[the remark this answers is not in the document]
Comment (NagyZoltanPeter) [reply apps/liteprotocoltester/monitoring/configuration/dashboards/nwaku-monitoring.json:91]: Ah, yes... create one, sure!

**Comment by NagyZoltanPeter** [comment]
> Great PR! Thanks for it! You are solving a very complex issue wrt how to properly track the "sent" messages 🥳
> 
> I'm not approving yet because we need to comment a little what is the best approach to track the "in" traffic.
> 
> If we merge as it is now we will get metrics of all the gross "in" traffic. In other words, we would track all the "in" traffic and even for the topics "my peer" are not interested in or messages that don't bypass the msg validations. I think this metric sounds interesting, to see how loaded "my peer" is, but I think that is something we can add in a separate PR.
> 
> On the other hand, I suggest having the "in" traffic metrics logic within the following `else` statement so that we can print the valid messages "my peer" is interested in:
> 
> https://github.com/waku-org/nwaku/blob/99149ea9dc252c8d6b6d563d0804c1c474a528db/waku/waku_relay/protocol.nim#L273-L274
> 
> As an additional context, the current approach is tracking the "in" traffic from: https://github.com/vacp2p/nim-libp2p/blob/d0af3fbe8559f69195657a360c3dd4ac4552c811/libp2p/protocols/pubsub/gossipsub.nim#L498
> 
> Whereas the suggested approach will track "in" traffic from: https://github.com/vacp2p/nim-libp2p/blob/d0af3fbe8559f69195657a360c3dd4ac4552c811/libp2p/protocols/pubsub/gossipsub.nim#L445

@Ivansete-status.
Yes, you see it very well, but all these was intentional. I played it with a lot before this solution.
The problem is, if we calculate only the filtered messages as `IN` we will lie to ourself about bandwidth stress on the node. So we may not find specific shard that spams us. 
This measurement has a follow up task, aka dynamic subscribe/unsubscribe to shards based on bandwidth requirements later on. 
But agree as a measurement we may differentiate between relay/shard/in and relay/shard/valid-in (which is valid in the terms we try to propagate it).

BTW, normally we should get messages on topics via relay we are subscribed to... if not that is the case we are spammed somehow.

What we already discussed with @gabrielmer that we need to prepare a libp2p PR that enables better fit hooks during the relay/publish process for getting known to possible errors come across.
Currently, we can know very little what went wrong.

**Comment by NagyZoltanPeter** [comment]
> Great PR! Thanks for it! You are solving a very complex issue wrt how to properly track the "sent" messages 🥳
> 
> I'm not approving yet because we need to comment a little what is the best approach to track the "in" traffic.
> 
> If we merge as it is now we will get metrics of all the gross "in" traffic. In other words, we would track all the "in" traffic and even for the topics "my peer" are not interested in or messages that don't bypass the msg validations. I think this metric sounds interesting, to see how loaded "my peer" is, but I think that is something we can add in a separate PR.
> 
> On the other hand, I suggest having the "in" traffic metrics logic within the following `else` statement so that we can print the valid messages "my peer" is interested in:
> 
> https://github.com/waku-org/nwaku/blob/99149ea9dc252c8d6b6d563d0804c1c474a528db/waku/waku_relay/protocol.nim#L273-L274
> 
> As an additional context, the current approach is tracking the "in" traffic from: https://github.com/vacp2p/nim-libp2p/blob/d0af3fbe8559f69195657a360c3dd4ac4552c811/libp2p/protocols/pubsub/gossipsub.nim#L498
> 
> Whereas the suggested approach will track "in" traffic from: https://github.com/vacp2p/nim-libp2p/blob/d0af3fbe8559f69195657a360c3dd4ac4552c811/libp2p/protocols/pubsub/gossipsub.nim#L445

@Ivansete-status.
Yes, you see it very well, but all these was intentional. I played it with a lot before this solution.
The problem is, if we calculate only the filtered messages as `IN` we will lie to ourself about bandwidth stress on the node. So we may not find specific shard that spams us. 
This measurement has a follow up task, aka dynamic subscribe/unsubscribe to shards based on bandwidth requirements later on. 
But agree as a measurement we may differentiate between relay/shard/in and relay/shard/valid-in (which is valid in the terms we try to propagate it).

BTW, normally we should get messages on topics via relay we are subscribed to... if not that is the case we are spammed somehow.

What we already discussed with @gabrielmer that we need to prepare a libp2p PR that enables better fit hooks during the relay/publish process for getting known to possible errors come across.
Currently, we can know very little what went wrong.

**Review by gabrielmer — APPROVED** [review]
Love it! Thanks so much! 😍 

if you can please attach a screenshot of how the Grafana dashboard looks like after the changes, as it's hard to review that.

Really like this solution :))

**Comment by Ivansete-status** [comment]
... 
> What we already discussed with @gabrielmer that we need to prepare a libp2p PR that enables better fit hooks during the relay/publish process for getting known to possible errors come across. Currently, we can know very little what went wrong.

Thanks for the explanation @NagyZoltanPeter! Fair enough, what you mention makes a lot of sense. Nevertheless, it is still relevant to also add statistics and metrics for the "net in" traffic so that we can have a ratio between the gross and the net traffic that we have. Maybe something to consider in upcoming PRs but that would be very relevant IMHO ;P

**Review by Ivansete-status — APPROVED** [review]

**Comment by NagyZoltanPeter** [comment]
> ...
> 
> > What we already discussed with @gabrielmer that we need to prepare a libp2p PR that enables better fit hooks during the relay/publish process for getting known to possible errors come across. Currently, we can know very little what went wrong.
> 
> Thanks for the explanation @NagyZoltanPeter! Fair enough, what you mention makes a lot of sense. Nevertheless, it is still relevant to also add statistics and metrics for the "net in" traffic so that we can have a ratio between the gross and the net traffic that we have. Maybe something to consider in upcoming PRs but that would be very relevant IMHO ;P

Sure, will extend in a separate PR. Thanks for it.

File: waku/waku_relay/protocol.nim [hunk 2]
```diff
@@ -150,6 +153,94 @@ proc initProtocolHandler(w: WakuRelay) =
...
+  proc decodeRpcMessageInfo(
+      peer: PubSubPeer, msg: Message
+  ): Result[
+      tuple[msgId: string, topic: string, wakuMessage: WakuMessage, msgSize: int], void
+  ] =
+    let msg_id = w.msgIdProvider(msg).valueOr:
+      warn "Error generating message id",
+        my_peer_id = w.switch.peerInfo.peerId,
+        from_peer_id = peer.peerId,
+        topic = msg.topic,
```
[the remark this answers is not in the document]
Comment (NagyZoltanPeter) [reply waku/waku_relay/protocol.nim:166]: Done

File: waku/waku_relay/protocol.nim [hunk 3]
```diff
@@ -150,6 +153,94 @@ proc initProtocolHandler(w: WakuRelay) =
...
+
+    waku_relay_network_bytes.inc(
+      msgSize.int64, labelValues = [pubsub_topic, if onRecv: "in" else: "out"]
+    )
+
+  proc onRecv(peer: PubSubPeer, msgs: var RPCMsg) =
+    for msg in msgs.messages:
+      let (msg_id_short, topic, wakuMessage, msgSize) = decodeRpcMessageInfo(peer, msg).valueOr:
+        continue
+      # logMessageInfo(peer, topic, msg_id_short, wakuMessage, onRecv = true)
```
[the remark this answers is not in the document]
Comment (NagyZoltanPeter) [reply waku/waku_relay/protocol.nim:229]: Added a better comment to explain the situation.

File: apps/liteprotocoltester/monitoring/configuration/dashboards/nwaku-monitoring.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (NagyZoltanPeter) [reply apps/liteprotocoltester/monitoring/configuration/dashboards/nwaku-monitoring.json:91]: Added one. This one is from waku-simulator, but nwaku-compose is very much the same except for one node not many.

File: waku/waku_relay/protocol.nim [hunk 4]
```diff
@@ -150,6 +153,94 @@ proc initProtocolHandler(w: WakuRelay) =
...
+
+    let msgSize = msg.data.len + msg.topic.len
+    return ok((msg_id_short, msg.topic, wakuMessage, msgSize))
+
+  proc logMessageInfo(
+      peer: PubSubPeer,
+      topic: string,
+      msg_id_short: string,
+      msg: WakuMessage,
+      onRecv: bool,
```
[the remark this answers is not in the document]
Comment (NagyZoltanPeter) [reply waku/waku_relay/protocol.nim]: Removed.

File: waku/waku_relay/protocol.nim [hunk 5]
```diff
@@ -150,6 +153,94 @@ proc initProtocolHandler(w: WakuRelay) =
...
+      return err()
+
+    let msg_id_short = shortLog(msg_id)
+
+    let wakuMessage = WakuMessage.decode(msg.data).valueOr:
+      warn "Error decoding to Waku Message",
+        my_peer_id = w.switch.peerInfo.peerId,
+        msg_id = msg_id_short,
+        from_peer_id = peer.peerId,
+        topic = msg.topic,
```
[the remark this answers is not in the document]
Comment (NagyZoltanPeter) [reply waku/waku_relay/protocol.nim]: Renamed.

**Pull request merged**